### PR TITLE
Integrate findlib into compiler toolchain

### PR DESCRIPTION
**Pull request by nojb**
This pull request integrates [`findlib`](http://projects.camlcity.org/projects/findlib.html) support directly into the compiler toolchain, considerably improving the user experience.

> While this patch is already functional, it is not yet complete.  I am posting this to request feedback, and gauge interest, from the wider community.

*Acknowledgement:* this project is only possible thanks to Gerd Stolpmann's work and in fact much of the code has been adapted from `ocamlfind`'s `Frontend` module.  Thanks!

## Example

```ocaml
$ cat > test.ml <<END
let f () =
  for%lwt i = 1 to int_of_string Sys.argv.(1) do
    Lwt_io.printlf "%d" i
  done

let () =
  Lwt_main.run (f ())
END
$ ocamlc -package lwt.ppx -package lwt.unix test.ml
$ ./a.out 2
1
2
```

Note that it is not necessary to pass `ocamlfind`'s `-linkpkg` flag.

## Description

For the moment only `ocaml{c,opt}` are adapted.  The treatment of `ocamlfind ocaml{c,opt}` flags is as follows:

- The new flag `-package` is added.  This flag corresponds to the homonym flag of `ocamlfind`.  Its effect is to add include directories for the named package (and all dependencies), the necessary archives when linking, the necessary `ppx` preprocessors, etc.

- The `findlib` packages are assumed to be installed at `$(ocamlc -where)/..`. This is the standard convention and the one used by `opam` as well.

- `-linkpkg` flag is not necessary, the compiler knows when it is linking and does the right thing.

- We behave as if `-dllpath-all` is always passed.  Is there any use-case where we do not want this to be the case?

- `-syntax`, `-ppopt`, `-ppxopt`: not yet implemented, should be easy to add them.

- `-predicates`, `-dontlink`, `dllpath-pkg`.  Have not exposed these at this stage. Would appreciate feedback on the use-cases of these options.

- `linkopts` entry in `META` file is not yet honored. This needs a little thought as it involves command-line parsing.

## Some implementation points

- The `findlib` project is added as a `git submodule`.  The idea is for `findlib` to continue to be developed independently and not add to the maintenance burden of the compiler.

- We use the public `Findlib` API to access the `META` database. A small module `Findlib_helper` is added to the `utils` directory for this purpose.  This is the sole extra dependency added to the compiler codebase.

- Currently there is a big-ish addition to the Makefile necessary to compile `findlib` with the bootstrap compiler.  Completely standard stuff, it can probably be elided by making some small changes to the upstream `findlib` Makefile.

- The `configure` flag `--with-findlib` activates `findlib` support.  While I expect this to become the default in "normal" installs, if this flag is not passed nothing is changed and everything compiles as before (in particular it is not necessary to checkout the submodule).  This part is a little rough and uses `cpp` for the conditional compilation, and will be improved.

- The whole patch adds less than 300 lines to the compiler codebase.

## Out of scope: installation, querying

This pull request does not try to integrate `ocamlfind install` feature. Installation is better handled by a package manager such as `opam` or some other means.

Similarly, no options are exposed to perform arbitrary queries of the `META` database.  For now, I am leaving this to the `ocamlfind` tool.

## Conclusion

Personally I feel this project could considerably improve the usability of the OCaml toolchain.  In my experience beginners are often surprised that one needs to install yet another tool and learn a new set of flags and settings simply to make use of a third-party library (maybe installed by `opam`). 
`ocamlbuild -use-ocamlfind` of course helps here, but 1) often one cannot/does not want to use it, and 2) ~~it does not work on Windows~~ `findlib` integration seems too basic to depend on the chosen build system.

We are all aware of the issues caused by the fragmentation of build and packaging solutions in the OCaml community.  The `ocamlfind` tool plays a very useful role in this setting as the de-facto standard for interoperability of third-party libraries.

If we are able to integrate this tool in the compiler toolchain in a non-invasive way, while at the same time improving usability, then this seems to me like a very worthwhile addition.

Any and all opinions warmly welcome.

/cc @gerdstolpmann /cc @alainfrisch /cc @dbuenzli

### CODE REVIEW

**Comment by gasche** [comment]
I find this a very exciting development, thanks! On the other hand, we are in the middle of the 4.05 release process, which means that I expect many maintainers or frequent contributors to be busy stabilizing things in preparation for the release. Don't feel bad if feedback seems to be lagging a bit (I won't have time to have a close look at this soon); and please don't distract people too much!

(I'm surprised by the mention that `ocamlbuild -use-ocamlfind` does not work on Windows, given that there seems to be Windows users of ocamlbuild. I'm not one of them and thus not in capacity to test issues/improvements, but any help there would be welcome. This is, of course, fairly orthogonal to the present PR.)

**Comment by nojb** [comment]
Hi @gasche, thanks for the comment!  Indeed, the timing might be a little inopportune.  It is just that I happened to have some hours to hack on this and could not resist the temptation of putting it out there for comments.  But of course I don't expect any substantial discussion before 4.05 is out the door.

Regarding `ocamlbuild` I think I was not as precise as I should have been: indeed the basic `ocamlbuild` does work on Windows (as far as I remember), just that there are issues regarding interaction with the underlying shell, etc.  And `findlib` support just seems to me a more basic concept that should not be dependent on a build system, `ocamlbuild` or other.

**Comment by adrien-n** [comment]
ocamlbuild -use-ocamlfind definitely works or at least /worked/ on Windows but that doesn't make this integration least interesting. :)

File: configure [hunk 1]
```diff
@@ -2184,6 +2189,11 @@ else
   else
   inf "        afl-fuzz always enabled .. no"
   fi
+  if test "$with_findlib" = "true"; then
```
Comment (adrien-n) [inline configure]: You can directly

    if "${with_findlib}"; then

But really that's very minor.

**Comment by nojb** [comment]
Hi @adrien-n, thanks for the comment! Yes, as I clarified above, the issue is not that `ocamlbuild` does not work in Windows (which it does) but rather the more general point that in my opinion `findlib` integration should not depend on the build system at all.  I will update the PR description with this clarification.

**Review by adrien-n — COMMENTED** [review]
There should probably be a way to configure that. 

I'm not entirely clear about Config.standard_library: it is influenced by environment variables?

I think some distributions à la NixOS might have an issue with the current approach. I'll see if I can get some input on that.

As far as I'm concerned, I've used the ability to get packages from a non-standard location several times. Including while the package was not /yet/ in its final location.

**Review by adrien-n — CHANGES_REQUESTED** [review]
This comments-out large chunks of the previous commit. It makes things needlessly difficult to follow and go through. =/

**Comment by nojb** [comment]
Yes, `Config.standard_library` is influenced by some environment variables (see here: https://github.com/ocaml/ocaml/blob/trunk/utils/config.mlp#L22-L29).

Thanks for the feedback on the need to be able to get packages from a non-standard location. I will think about a good approach to do this.  `findlib` itself allows to override this path using the environment variable `OCAMLPATH` (but I need to check if this is working in this code).  In any case we might want to add an explicit command-line option as well.

**Comment by nojb** [comment]
Yes @adrien-n, I have not rebased/cleaned up the history so it is not very convenient to examine the individual commits.  I do not think it makes sense to do this yet as I am expecting some amount of back-and-forth. Having said that, I think the patch is small enough that it is possible to read it all at once.

**Comment by dbuenzli** [comment]
Personally I'm not convinced this will lead to a simpler build eco-system. 

Most of what I write in `META` files is already written in the file system (see e.g. [`odig`](https://github.com/dbuenzli/odig)'s automatic [toplevel loader](https://github.com/dbuenzli/odig/issues/10) or `META` [files generation](https://github.com/dbuenzli/odig/blob/master/toys/metagen.ml) from build artefacts for a constructive proof of this). Besides for most packages ocamlfind package names map directly to archive names; there are too many names and notions of packages in the eco-system and the notion of ocamlfind name seems to be one we could eschew in the future. 

As such I'm not sure that standardizing findlib now is what will lead us to a simpler eco-system in the long term.

Also IIRC some of @lpw25's idea about namespaces may subsume part of `findlib`'s functionality.

**Comment by lefessan** [comment]
Although I share the concern to help beginners with OCaml, I don't think the solution is to put `findlib` in the compiler. 

What about using the new compiler plugins architecture ? If we simplify the way plugins are searched for, it would give something like:
```
ocamlc -plugin findlib -package lwt.ppx -package lwt.unix test.ml
```
(with no namespace on plugin arguments, which I think it the best option)

**Comment by gerdstolpmann** [comment]
First of all, I think it is great when more people think about how library management should be done. Let me add a few comments:

 - If the goal is to hide ocamlfind while exposing its functionality more directly, I think it is a good goal. Library management is something very technical, and while the details should be still visible to those who want (or need) to know, it would definitely improve the user experience.
 - Almost nothing comes for free though, and I think in this case the price would be that we have a split implementation of some part functionality (in particular what's in findlib's frontend.ml which is repeated here in `Findlib_helper`), and an additional dependency from ocaml to findlib (though only in terms of a git submodule). Also, some special features cannot work anymore, like toolchains and customized configurations where the compiler invoked by ocamlfind isn't the one that is found in PATH.
 - The problem with the stuff in frontend.ml can be tackled by changing the findlib code base so that the function that juggles the command-line arguments passed to the compiler can be called from outside. That only takes some refactoring.
 - A bigger problem is the new dependency between ocaml and findlib. With the "git submodule" method we create a tight binding of the compiler to a very specific version of findlib. You cannot update the findlib functionality independently. Note that this ability was already very helpful in the past, e.g. for the introduction of `Bytes` where findlib was used to distribute a new `bytes` library also for already released versions of ocaml.
 - Toolchains are increasingly used for cross-compiling. Here the separation between findlib and ocaml is really a prerequisite. Findlib is here in the role of picking the right compiler among several installed versions. Toolchains would no longer work if the indirection via `ocamlfind` was no longer the common way of driving the compiler.

The last two points are the difficult ones. Fabrice's idea of using plugins could help loosing the tightness of the new dependency (but details still need to be worked out). I don't see how we could save toolchains, though.

Let me also add some very general thoughts. I mean the problem is the layered compiler driver: the core is directly part of the compiler, and ocamlfind wraps that driver in the simplest possible way to add its functionality. If we want better/cooler/hotter drivers, let's think about different ways of arranging the code. For example, there could be a unified `ocamluc` driver that doesn't include the compiler, but that loads plugins to get certain tasks done. In other words: the compilers and findlib would be plugins of this unified driver. This way you can achieve a number of cool things, e.g. compile bytecode and native code together, or control the compiler via a socket, or integrate automatic preprocessing.

**Comment by nojb** [comment]
Hi @lefessan, thanks for the comment!  I am not familiar with the new plugin architecture.  I will look into it before replying more fully.

**Comment by nojb** [comment]
Hi @gerdstolpmann, thank you for your detailed comments, they are very appreciated.

As you write in your point 3, the duplication of code between `Frontend` and `Findlib_helper` can be avoided in a large part by refactoring/pulling some more code into the `findlib` library.  This is what I had in mind from the beginning.

I need to look into the `-toolchain` option because I am not very familiar with it, but as far as cross-compiling is concerned, I **think** there are plans for OCaml to do this natively (@shindere may be able to say something about this), so in my mind this is not such a big issue.

On the other hand, I would like to understand better the role `ocamlfind` played in the introduction of the `bytes` functionality.  Do you have a link to the relevant discussion at the time ?

As a general point, while I can see the technical merits of both a `plugin`-based approach, as suggested by @lefessan, or a `driver`-based approach as suggested by you, I do feel that in both cases the user experience is affected by introducing more intermediate, extraneous, "stuff".  Removing this "stuff" is the motivation of this project, and so far I do not feel convinced by the arguments against the present approach.  Of course I need to study them better to have a more informed opinion.

Thanks!

**Comment by nojb** [comment]
Hi @dbuenzli, thanks for your comments!

I hear your point about short-term vs long-term goals.  On the other hand, a `ocamlfind`-based library ecosystem is what we have **today**.  If we have something better in the future (even in the near future) then we will adapt to that when it happens, but I feel like not doing anything to improve the present situation because of possible future improvements is not such a great idea either.

**Comment by adrien-n** [comment]
I don't believe that these (plugins or driver) will necessarily add
layers for the users. It's a matter of default behaviour.

Fabrice gave an example with an explicit command-line switch but we
could also decide that if the plugin is present, it is loaded by
default. The motivation is that if you've installed it, you probably
want to use it and if you don't want to use the plugin, you're not going
to install it.

There are a few things to keep in mind with respect to compatibility of
plugins or of a driver. IIRC Debian has some platforms without dynamic
loading (but that should be verified) and might prefer keeping the
current approach.

PS: I like the idea of a driver because the socket control would enable
a lot of new functionalities but I don't know how big a step that would
be right now: smaller steps might prove more efficient.

**Comment by dbuenzli** [comment]
> On the other hand, a ocamlfind-based library ecosystem is what we have today

Sure but as can be shown what we have *today* is absurd. I see no point in enshrining absurd technological solutions.

**Comment by lpw25** [comment]
I agree with @dbuenzli. The current situation for managing libraries in OCaml is pretty bad, and merging parts of it into the compiler will mean we are stuck with it forever.

**Comment by nojb** [comment]
Just for the sake of concreteness, it would be great if @dbuenzli and/or @lpw25 could list the reasons why the current situation for managing libraries is "pretty bad" or "absurd".

**Comment by adrien-n** [comment]
As a first step, what would be the names you would want to settle for in
the long run?

AS far as I'm concerned, I see the ocamlfind package names as the ones
that would remain stable even if not handled by ocamlfind anymore.
Therefore I don't see the issue with using them in the compiler.

I'm really geniously curious about what you have in mind and with which
timeframe.

**Comment by lefessan** [comment]
For example, in some other languages, access to libraries is done in the language itself, through the use of namespaces. I did such a proposal for OCaml a long time ago. Namespaces solve both the problem solved by `findlib`, plus the problem of conflicts between module names.

**Comment by lpw25** [comment]
@nojb Quite a few things are bad, but I think a particularly illustrative example is how many separate names a library has in OCaml:

1. The name of its main module.
2. The name of its `cma` file.
3. The name of its library in ocamlfind
4. The name of its pacakge on opam

This illustrates that the current system has far too many layers. These layers also mean that you have to specify to same information (modules in the library, dependencies, ...) in multiple different places (opam file, META file, .merlin file, build system, top-most .ml file, ...).

**Comment by gerdstolpmann** [comment]
@lpw25 I don't understand this. An opam package may have several libraries consisting of multiple cma's each, with tons of modules in it. Containers in containers. Of course, on every level there is separate naming. Sure, there is the simplistic case where you put one module into one cma which goes into one library, distributed as a package of its own, and it might feel absurd then.

When you ask me which naming levels can go away:
 - OPAM could decide to only distribute one library per package. This is imaginable, but definitely a challenge for the build systems. Or, OPAM keeps multiple libraries, but the package name is taken as the set of library names, and becomes then a purely technical name the user normally does not see.
 - The name of the cma does not mean anything. It could be a random number generated by the build system, and that is normally hidden from the user.

The findlib name is not in this list because it is the entity to which we attach dependency information.

Other languages have different absurdities. E.g. in Java, you create deep directory hierarchies, only for naming. @lefessan Java is a good example that namespaces do not solve the problem. Look at the complexity added by maven. Despite namespaces there is a separate Maven name! It's like a curse: names require name resolution, and another level of indirection is always useful in some cases.

**Comment by gerdstolpmann** [comment]
@nojb Regarding bytes, you might have noticed that there is an empty library `bytes`. It is not always empty. When you build findlib for an older OCaml version, it contains an emulation of the Bytes module (roughly as alias of String). So, when you link with `-package bytes` you can be sure that there is a Bytes module no matter which version of OCaml you have.

This was thought as a helper to accelerate the porting to Bytes, in particular for code bases that need to keep compatibility with older OCaml releases.

Technically, we just enhanced older OCaml releases after their release.

**Comment by nojb** [comment]
Thanks for the explanation @gerdstolpmann. Indeed, this maneuver does not seem easy to do without a separate driver program.

**Comment by lpw25** [comment]
@gerdstolpmann In most cases libraries contain a single cma, which exposes either a single packed module or a module containing module aliases to hidden submodules. So the current situation is absurd for most cases not just libraries of a single module.

I have a namespacing proposal being actively worked on, and that I've presented to the rest of the development team in the past which removes the need for cma files, and creates the top-level module automatically from the directory structure (which is generally controlled by opam). That effectively leaves us with just the opam and ocamlfind names.

In the common case it also removes the need for ocamlfind leaving us with a single name.
Now ocamlfind is a powerful tool, and not all its facilities and use cases are covered by the initial proposal, but in the long run I intend to gradually decrease the need for it. This would obviously be hampered by including it within the compiler itself.

**Comment by gerdstolpmann** [comment]
@lpw25 Well, when I look at the real software basis, I don't see that many packed modules. Because it is complicated and another layer that is not that well integrated. The reality is rather that most libraries are collections of toplevel modules, managed as cmas. Still in 2017. So, if your suggestion eases the creation of packs and removes this hard-to-understand distinction between packs and aliases, I'm all for it.

Nevertheless, you need a naming level where you express the library dependencies. I guess this is not also already built in? My point is only that namespaces do not solve all problems.

Btw, I agree that findlib shouldn't be part of the compiler, but for different reasons.

**Comment by lpw25** [comment]
> So, if your suggestion eases the creation of packs and removes this hard-to-understand distinction between packs and aliases, I'm all for it.

That is indeed it's primary purpose.

> Nevertheless, you need a naming level where you express the library dependencies. I guess this is not also already built in? My point is only that namespaces do not solve all problems.

Well the dependencies needed to build a package are already listed in the opam file. So if the compiler already knows where to find the `Core` module (`.opam/lib/core/`) then there is no need to tell your build system or the compiler that you depend on `Core` and so you avoid the need to state your dependencies twice.

This somewhat relies on the other parts of the proposal causing libraries to be within a (auto-generated) top-level module, which makes things easier because the compiler is always looking for something like `Core.Foo` rather than looking for a `Foo` without knowing which library it comes from.

As I said, this is not enough to handle all of ocamlfinds use cases in terms of managing dependencies, but it should handle common cases. It also aims to integrate easily with ocamlfind, so that people can get the easy pack/alias stuff without having to stop using ocamlfind if they don't want to.

**Comment by dbuenzli** [comment]
> Just for the sake of concreteness, it would be great if @dbuenzli and/or @lpw25 could list the reasons why the current situation for managing libraries is "pretty bad" or "absurd".

I have already answered this question. Simply do a pin of the `odig` project and try to play with the features I mentioned. I will add to those `odig guess-deps` and the fact that `odig`'s toplevel helpers are able to load the deps of a single compilation unit without having to lookup *any* metadata file. By hijacking the undefined symbol error paths of the toplevel we could actually load libraries on demand which would kill the needs for most `.ocamlinit` files and this mostly without using any metadata (some is needed in certain cases, e.g. link-time selection of implementation of an interface, see below).

To repeat myself, most of what we write in `META` files can be inferred by looking at what packages wrote on the file system and most ocamlfind packages names are one-to-one with library archive names. I'd say this covers 97% of packages, a few of them can be converted to get to a 99%.

For the 1% remaining cases ( e.g. when linking order is implicit matters due to side effects tricks, or ambiguous resolutions exist because of more than one module implements the same `-opaque` interface) we could go with a few standardized `opam` v2.0 extension fields which compilation helpers could lookup to guide the process. 

That way we can gradually get to a system where 1) We mostly don't have to specify anything and hence we are never out of sync since dependencies are inferred from the build artefacts themselves  2) We don't have two metadata file formats 3) We try to kill a few names along the way; I don't really mind which ones are killed, if it's the cmas or the ocamlfind ones since as I said those are mostly one to one, modulo a few replacement of `_` or `-` by `.`.

There is absolutely no way we can drop `ocamlfind` now or in the near future, it is deeply connected to current build system in use. Using the ideas above pervasively will take some time but both systems can perfectly live harmoniously. If people want to move in a world that kills the need to replicate what can be inferred via a data-driven approach, standardizing `ocamlfind` now in the compiler will certainly not make this easier. 

Regarding `bytes` I actually think that the way this was solved rather obscured things than enlighted them. There's no reason why this had to be tuck in `ocamlfind` when all other similar additions were done via a compatibility package (e.g. `result`, `uchar`) that live independently.

**Comment by nojb** [comment]
Thanks @dbuenzli for the explanation.

Still, here is another way to look at this project:

All that is proposed is to add a flag to the compiler, `-package PKG`, where `PKG` denotes a library and adds all necessary compiler flags to compile with `PKG` (includes, archive names, `ppx` preprocessor, etc).

Today, we must use `findlib` names for `PKG`.  Tomorrow we could be using some `opam` name, namespaces or whathaveyou.  In fact one can imagine implementing some type of "cascade" logic
to support the migration from `ocamlfind` to whatever comes next.  From this point of view the proposed flag could even **help** the migration to a better system by providing a stable interface.

In short, as long as we do not expose too many aspects of `findlib`, then it can remain essentially an implementation detail.  The use of `findlib` names does not preclude using some other class of names later on, with user-defined priorities if needed, etc.

**Comment by dbuenzli** [comment]
I don't see *any* benefit. 

The current system is cumbersome but it works now. Every build tool out there knows how to use `ocamlfind`. So basically you would like these tools to migrate to use the compiler's `-package` flag so that they do the same thing as `ocamlfind` now. However if later it turns out that the `-package` flags we need in the compiler have a slightly different semantics then we are bust. 

This endeavour only seems a good way of introducing more entropy and technical debt in the system.

**Comment by alainfrisch** [comment]
What about making OCaml library files (.cma/.cmxa) more powerful?  They already include some link-time options.  They could as well embark some compilation-time flags (-I, -ppx, etc...) and even refer to other .cma/.cmxa files.  One would simply need to refer to a .cma file at compile-time to get all required flags implied by it.  Assuming that packages are installed in a standard-enough location (which would be the case if installation is done through OPAM), using a library would become as easy as:

  ocamlc -c +package1/lib1.cma +package2/lib2.cma foo.ml

**Comment by nojb** [comment]
I like this approach, but one issue I can see right away is that there are concepts which are not directly tied to cmas, for example `ppx` preprocessors.

**Comment by alainfrisch** [comment]
> I like this approach, but one issue I can see right away is that there are concepts which are not directly tied to cmas, for example ppx preprocessors

If we include -ppx directives in .cma file, this concept becomes tied to cmas :)  One might need to deal specially with .cmxa file with no object files (I think this creates problem with linkers on some OS), but this is a detail.

**Comment by xavierleroy** [comment]
This is a textbook example of a completely premature pull request: there is a very interesting (and deep) design discussion that should take place more publically (e.g. on caml-list@inria.fr), but until we have some agreement on what to do, the proof-of-concept implementation isn't useful.  So, I suggest to move the discussion to caml-list and close this PR.

My down-to-earth take on the question:
* The party line these days is to take stuff away from the core compiler distribution and turn it into independent libraries or tools.  So, the opposite direction needs very strong arguments indeed.
* I agree that `ocamlfind` works very well as a separate tool for the tasks it does today.  There isn't much to gain by integrating as is.  Build systems such as `ocamlbuild` know when to add `ocamlfind` to the command-line when needed.
* I also agree that adding some notions of packages and namespaces to the OCaml compilers and perhaps even to the language could result in a better user experience than `ocamlfind`.  But there's a lot of design work that remains to be done.

**Comment by nojb** [comment]
Thanks for the feedback @xavierleroy.
